crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,495 @@
+            let last_item_line_offset = locator.line_end(last_item.end());
+            if last_item_line_offset == locator.line_end(self.end()) {
+                (false, last_item.end())
+            } else {
+                (true, last_item_line_offset)
+            }
+        };
+        let mut prelude = locator
+            .slice(TextRange::new(self.start(), prelude_end))
+            .to_string();
+        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));
+
+        let joined_items = if self.multiline {
+            let indentation = stylist.indentation();
+            let newline = stylist.line_ending().as_str();
+            prelude = format!("{}{}", prelude.trim_end(), newline);
+            join_multiline_dunder_all_items(
+                &sorted_items,
+                locator,
+                parent,
+                indentation,
+                newline,
+                needs_trailing_comma,
+            )
+        } else {
+            Some(join_singleline_dunder_all_items(&sorted_items, locator))
+        };
+
+        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
+        SortedDunderAll::Sorted(new_dunder_all)
+    }
+}
+
+impl Ranged for DunderAllValue {
+    fn range(&self) -> TextRange {
+        self.range
+    }
+}
+
+#[derive(Debug)]
+enum SortedDunderAll {
+    AlreadySorted,
+    Sorted(Option<String>),
+}
+
+fn collect_dunder_all_lines(range: TextRange, locator: &Locator) -> Option<Vec<DunderAllLine>> {
+    // Collect data on each line of `__all__`.
+    // Return `None` if `__all__` appears to be invalid,
+    // or if it's an edge case we don't care about.
+    let mut parentheses_open = false;
+    let mut lines = vec![];
+    let mut items_in_line = vec![];
+    let mut comment_in_line = None;
+    for pair in lexer::lex(locator.slice(range).trim(), Mode::Expression) {
+        let (tok, subrange) = pair.ok()?;
+        match tok {
+            Tok::Lpar | Tok::Lsqb => {
+                if parentheses_open {
+                    return None;
+                }
+                parentheses_open = true;
+            }
+            Tok::Rpar | Tok::Rsqb | Tok::Newline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                }
+                break;
+            }
+            Tok::NonLogicalNewline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                    items_in_line.clear();
+                    comment_in_line = None;
```

Comment:
You might be able to avoid cloning the value inside of `DunderAllLine` by using `std::mem::take`. `std::mem::take` *takes* the value from `items_in_line` and replaces it with the `Default` (an empty list in that case). 
```suggestion
                    lines.push(DunderAllLine::new(
                        std::mem::take(&mut items_in_line),
                        comment_in_line,
                        range.start(),
                    ));
                    comment_in_line = None;
```